nav2_route/include/nav2_route/breadth_first_search.hpp

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2020, Samsung Research America
+// Copyright (c) 2023 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_ROUTE__BREADTH_FIRST_SEARCH_HPP_
+#define NAV2_ROUTE__BREADTH_FIRST_SEARCH_HPP_
+
+#include <queue>
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+#include "nav2_costmap_2d/cost_values.hpp"
+#include "nav2_costmap_2d/costmap_2d.hpp"
+
+namespace nav2_route
+{
+
+/**
+ * @struct nav2_route::SimpleNode 
+ * @brief A Node implementation for the breadth first search
+ */
+struct SimpleNode
+{
+  SimpleNode(unsigned int index)
+  : index(index), 
+    explored(false) {}
+
+  unsigned int index;
+  bool explored;
+};
+
+/** 
+ * @class nav2_route::BreadthFirstSearch
+ * @brief Preforms a breadth first search between the start and goal
+ */ 
+class BreadthFirstSearch
+{
+public:
+  typedef SimpleNode * NodePtr;
+  typedef std::vector<NodePtr> NodeVector;
+
+  /**
+   * @brief Set the costmap to use 
+   * @param costmap Costmap to use to check for state validity
+   */
+  void setCostmap(nav2_costmap_2d::Costmap2D * costmap);
+
+  /**
+   * @brief Set the start of the breadth first search
+   * @param mx The x map coordinate of the start
+   * @param my The y map coordinate of the start 
+   */ 
+  void setStart(unsigned int mx, unsigned int my);
+
+  /**
+   * @brief Set the goal of the breadth first search
+   * @param mxs The array x map coordinate of the goals
+   * @param my The array y map coordinate of the goals
+   */ 
+  void setGoals(std::vector<unsigned int> mxs, std::vector<unsigned int> mys);
+
+  /**
+   * @brief Find the closest goal to the start given a costmap, start and goal
+   * @param closest_goal The coordinates of the closest goal
+   * @return True if the search was successful
+   */ 
+  bool search(unsigned int & closest_goal_index);
```

Comment:
I forget off the top of my head what the algorithmic failure modes I use across this project. If I do the whole "return true/false to handle failures" thing, then this is perfect no problems.

If I do more of a "I assume you succeed to return the results in the return value and throw exceptions otherwise" thing, then we should do that here too. I imagine in the use of this in the system, if `false` is returned, we would throw a failure exception. But I forget how low level I introduce exceptions in this project vs keeping them at medium-level software to keep the pure algorithms exception free.

Reply:
You are using the same flow as the rest of the servers. (try catch list) 

All the errors that I could throw are the same as the planners. Would using those be alright?

Reply:
Check where they're caught, but I suppose that would be fine unless that would get caught somewhere separate. Just make sure the string of the exception makes clear its from here and not there.